log4j-core/src/main/java/org/apache/logging/log4j/core/config/builder/impl/DefaultComponentBuilder.java

```diff
@@ -17,104 +17,217 @@
+     * @throws NullPointerException if either {@code builder} or {@code type} argument is null
+     */
+    public DefaultComponentBuilder(final CB builder, final String pluginType) {
+        this(builder, pluginType, null, null);
     }
 
-    public DefaultComponentBuilder(final CB builder, final String name, final String type) {
-        this(builder, name, type, null);
+    /**
+     * Constructs a new instancer with the given configuration builder, plugin-type, name, and {@code null} value.
+     * @param builder the configuration builder
+     * @param pluginType the plugin-type of the component being built
+     * @param name the component name
+     * @throws NullPointerException if either {@code builder} or {@code type} argument is null
+     */
+    public DefaultComponentBuilder(final CB builder, final String pluginType, final @Nullable String name) {
+        this(builder, pluginType, name, null);
     }
 
-    public DefaultComponentBuilder(final CB builder, final String name, final String type, final String value) {
-        this.type = type;
-        this.builder = builder;
+    /**
+     * Constructs a new instance with the given configuration builder, plugin-type, name and value.
+     * @param builder the configuration builder
+     * @param pluginType the type (plugin-type) of the component being built
+     * @param name the component name
+     * @param value the component value
+     * @throws NullPointerException if either {@code builder} or {@code type} argument is null
+     */
+    public DefaultComponentBuilder(
+            final CB builder, final String pluginType, final @Nullable String name, final @Nullable String value) {
+        super();
+        this.builder = Objects.requireNonNull(builder, "The 'builder' argument must not be null.");
+        this.pluginType = Objects.requireNonNull(pluginType, "The 'type' argument must not be null.");
         this.name = name;
         this.value = value;
     }
 
+    /** {@inheritDoc} */
     @Override
-    public T addAttribute(final String key, final boolean value) {
-        return put(key, Boolean.toString(value));
+    public Component build() {
+        final Component component = new Component(pluginType, name, value);
+        component.getAttributes().putAll(attributes);
+        component.getComponents().addAll(components);
+        return component;
     }
 
+    /** {@inheritDoc} */
     @Override
-    public T addAttribute(final String key, final Enum<?> value) {
-        return put(key, value.name());
+    public T addComponent(final ComponentBuilder<?> builder) {
+        Objects.requireNonNull(builder, "The 'builder' argument must not be null.");
+        components.add(builder.build());
+        return self();
     }
 
+    /** {@inheritDoc} */
     @Override
-    public T addAttribute(final String key, final int value) {
-        return put(key, Integer.toString(value));
+    public @NonNull CB getBuilder() {
```

Comment:
```suggestion
    public CB getBuilder() {
```

Isn't `@NonNull` the default?